[benchmark] Added Benchmark Check Report

palimondo opened the pull request:
Integrates the `Benchmark_Driver check` command into `run_smoke_bench`, producing Markdown formatted report table, analyzing the quality of **newly added** benchmarks. 

Example output:

✅  | Benchmark Check Report
---|---
⛔️🔤 | `String.insert.EmojiChar.NearEnd` name doesn`t conform to UpperCamelCase convention. <br><sub> See http://bit.ly/UpperCamelCase
⚠️⏱ | `String.insert.EmojiChar.NearEnd` execution took at least 1871 μs. <br><sub> Decrease the workload of `String.insert.EmojiChar.NearEnd` by a factor of 2 (10), to be less than 1000 μs.
⛔️🔤 | `String.insert.EmojiChar.EndIndex` name doesn`t conform to UpperCamelCase convention. <br><sub> See http://bit.ly/UpperCamelCase
⚠️Ⓜ️ | `String.insert.EmojiChar.EndIndex` has very wide range of memory used between independent, repeated measurements. <br><sub> `String.insert.EmojiChar.EndIndex` mem_pages [i1, i2]:  min=[12, 22] 𝚫=10 R=[15, 18]

If there are no issue with added benchmarks, the result looks like this:

✅  | Benchmark Check Report
---|---

palimondo committed:
[benchmark] Doctor: emit mem_page details info

Promoting previously DEBUG message to INFO.

palimondo committed:
[benchmark] MarkdownReportHandler

`logging.Handler` that creates nicely formatted report from `BecnhmarkDoctor`’s `check` in Markdown table for display on GitHub.

palimondo committed:
[benchmark] Added Benchmark Check Report

Produce Markdown formatted report, analyzing the quality of newly added benchmarks.

eeckstein commented on benchmark/scripts/run_smoke_bench:
```diff
@@ -119,6 +119,8 @@ def test_opt_levels(args):
                             args.platform, output_file):
             changes = True
 
+    check_added(args, output_file)
```
This is now called by default - which is fine.
So the -check-added option is obsolete, right?
Can you replace -check-added with a -no-check-added option which disables the checking? This would be useful for running the script locally to check compiler changes (and not benchmark changes).

palimondo replied:
Makes sense. I'm on it...

eeckstein reviewed (changes requested):
Basically LGTM! (Please see my comment inline)

Also, can you add a temporary test-commit to this PR which introduces some new "bad" benchmarks, which violate several rules? So that we can test this change with "swift-ci benchmark".

palimondo committed:
[benchmark] run_smoke_bench -skip-check-added

Removed the option to `-check-added`, as it’s now run by default. Replaced with option to skip checking added benchmarks: `-skip-check-added`.

palimondo commented:
@eeckstein I've added a `-skip-check-added` option. We can test with #20837, that renames `InsertCharacter` benchmarks to `String.insert` applying the proposed naming convention, because current naming rule will complain the new names don't conform to `UpperCamelCase`.

eeckstein commented on benchmark/scripts/run_smoke_bench:
```diff
@@ -80,10 +83,6 @@ def main():
-        '-check-added', action='store_const',
-        help="Run BenchmarkDoctor's check on newly added benchmarks",
-        const=lambda args: check_added(args), dest='func')
     argparser.set_defaults(func=test_opt_levels)
```
Can you please remove "func" and call test_opt_levels directly.
We don't need it anymore and it will simplify the code.

palimondo replied:
Done.

palimondo committed:
[benchmark] Call test_opt_levels directly

Simplify the code by removing indirection:
After removal of -check-added option, only single function remains.

palimondo commented:
@eeckstein Removed the indirection, `test_opt_levels` is now called directly.

eeckstein commented:
Can you just temporarily add a test-commit to this PR, and then after the benchmark run, remove it again?

eeckstein commented:
@swift-ci benchmark

palimondo committed:
[benchmark] Don’t emit report for no added tests

eeckstein commented:
@swift-ci benchmark

eeckstein commented:
@swift-ci benchmark

palimondo commented:
I've noticed the MarkdownReportHandler always prints the header, which is useful when there are no validation issues, but it did so even when there were no added benchmarks. a6f26e2 works around that by not adding the handler when `added` set is empty. @eeckstein Do you want to re-run benchmarks again with that change, too?

palimondo commented:
Assuming all goes well, I've removed the DNM commit and if you agree the cd47f32 is trivial, once all passes test this can land while I take my 💤s… (it's midnight here) 😴

eeckstein commented:
@swift-ci smoke test

eeckstein commented:
@swift-ci smoke test

eeckstein commented:
@swift-ci benchmark

palimondo commented:
@milseman Can you help me understand the unusual memory variance seen in the demonstration? Similar variance issues manifested themselves over in [Breadcrumbs](https://github.com/apple/swift/pull/20769#issuecomment-441812500). Granted, "constant memory use" rule is quite unproven recommendation at this time: I [asked for advice in the forums](https://forums.swift.org/t/towards-robust-performance-measurement/11490/23) before…

milseman commented:
That's weird, let's make sure it's not a formal memory leak. Which one seems to be the biggest offender?

palimondo commented:
String.insert.EmojiChar.NearEnd mem_pages [i1, i2]: min=[11, 12] 𝚫=1 R=[0, 59]

Explanation of this metric: `i1` (measured with `--num-iters=1`), `i2` means `--num-iters=2`. There were 10 independent runs, 5 `i1`, 5 `i2`. The lowest memory used was 11 pages (11 * 4096B = 44kB) and it was like that for all the runs, because the first R (range) is 0. But for i2, the second R means the highest memory consumption seen was 12 + 59 = 71 pages (284 kB).

Looking at all of the above ⚠️Ⓜ️ gives me very 🐟y vibes...

milseman commented:
This might be because those benchmarks don't reset their state after every iteration:

```swift
@inline(__always)
func insertTowardsEndIndex(_ c: Character, in string: String, count: Int) {
  var workload = string
  var index = workload.endIndex
  for i in 0..<count {
    workload.insert(identity(c), at: index)
    if i % 1000 == 0 {
      index = workload.endIndex
    }
  }
  blackHole(workload)
}
@inline(never)
func run_InsertCharacterTowardsEndIndex(_ N: Int) {
  insertTowardsEndIndex("s", in: str, count: N * 3000)
}

```

It seems like `run_InsertCharacterTowardsEndIndex` should have an outer loop from `0..<N` and then run `insertTowardsEndIndex` with a count of exactly 3000. That would give constant size and linear perf, instead of what I assume is currently linear size and quadratic perf.

palimondo commented:
@milseman That's actually good to hear, we can fix that!

And looking at new [`Breadcrumbs` tests](https://github.com/apple/swift/pull/20769), would you say the variance is normal, or is these some kind of issue, too? These are my local results:
````
'Breadcrumbs.CopyUTF16CodeUnits.ASCII' mem_pages [i1, i2]: min=[22, 23] 𝚫=1 R=[19, 18]
'Breadcrumbs.IdxToUTF16.longMixed' mem_pages [i1, i2]: min=[300, 307] 𝚫=7 R=[15, 24]
'Breadcrumbs.IdxToUTF16Range.longASCII' mem_pages [i1, i2]: min=[1050, 1049] 𝚫=1 R=[14, 19]
'Breadcrumbs.IdxToUTF16Range.longMixed' mem_pages [i1, i2]: min=[310, 310] 𝚫=0 R=[28, 29]
'Breadcrumbs.UTF16ToIdx.longASCII' mem_pages [i1, i2]: min=[1089, 1091] 𝚫=2 R=[31, 28]
'Breadcrumbs.UTF16ToIdx.longMixed' mem_pages [i1, i2]: min=[300, 307] 𝚫=7 R=[26, 20]
'Breadcrumbs.UTF16ToIdxRange.longASCII' mem_pages [i1, i2]: min=[1102, 1089] 𝚫=13 R=[25, 36]
'Breadcrumbs.UTF16ToIdxRange.longMixed' mem_pages [i1, i2]: min=[305, 305] 𝚫=0 R=[14, 21]

palimondo commented:
Well, actually, what you were saying about `run_InsertCharacterTowardsEndIndex`, is the memory varied between runs because you presumed different `N` between the runs, right? The the way `BenchmarkDoctor` does this measurement is to first calibrate number of samples to run for ~1s and then do the 5 `i1` and 5 runs `i2` runs with the constant number of samples (half in the i2 case). There should no longer be any variance from auto-scaling of the benchmarks present.

milseman commented:
I was assuming "iteration" means the count that's passed into the run function, so `run_InsertCharacterTowardsEndIndex`'s `N` parameter is its "iteration". `run_InsertCharacterTowardsEndIndex` inserts `3000 * N` characters into a string, so it's memory usage would scale linearly with `N`.

Or, is that what you mean by "sample"? I thought multiple "samples" would be running the same benchmark on the same `N` multiple times.

palimondo commented:
You are correct. The `--num-iters` parameter becomes the `N` that's passed to `run_` functions. Which means these benchmarks need to be corrected to do constant workload regardless of `N`. I'll fix that. 

That issue wasn't caught by the [corresponding heuristic in `BenchmarkDoctor`](https://github.com/apple/swift/blob/92cf40dcd39b55242f8a9b4853f2eb180409f193/benchmark/scripts/Benchmark_Driver#L458), because it's masked by the relatively large intra `i-series` variance in memory used:
````python
normal_range = 15  # pages
…
if abs(min_i1 - min_i2) > max(range_i1, range_i2, normal_range):
````

Let's have a look at this one:
````
String.insert.EmojiChar.EndIndex mem_pages [i1, i2]: min=[11, 12] 𝚫=1 R=[37, 29]
````
The runs within each i-series, i.e. `i1` and `i2`, are executed with the exact same N count (1, 2). The minimum number of pages used was 11 and 12 respectively, but the variance within each series is **triple that**! Any idea what might be causing that?

milseman commented:
Is mem_pages a record of the number of pages used during overall execution, or peak memory usage? When the string grows in size, it can trigger an allocation and copy, so we'd expect to see many pages allocated and deallocated over the course of this benchmark.

palimondo commented:
The `mem_pages` is a peak as reported by `rusage`. We [record baseline](https://github.com/apple/swift/blob/0bdd3ef27566fc780c084084c80524ad95d60a22/benchmark/utils/DriverUtils.swift#L339) before we start running and another one when we're [done with the benchmark](https://github.com/apple/swift/blob/0bdd3ef27566fc780c084084c80524ad95d60a22/benchmark/utils/DriverUtils.swift#L533).

Thinking about this more… the ideal case assumes the allocator will reuse of same memory for all samples, which might not be the case, as we are still running a lot of samples and the same page might get allocated to different process, right? I'm not familiar with how the system's allocator does that in detail.

palimondo commented:
Hmm... that probably means that memory usage measurement needs to be always performed with a single sample only and therefore separated from the runtime measurements. What do you think?

palimondo commented:
Hmm… `./Benchmark_O --num-iters=1 --num-samples=1 --memory InsertCharacterTowardsEndIndexNonASCII --verbose | tail -n 6 | head -n 1` again...
10, 10, 11, 15, 11, 10, 21, 10, 10, 10, 19, 10, 20, 12
